File: C++/Infix To Prefix/InfixToPrefix.cpp
#include <iostream>
#include <stack>
#include <algorithm>

using namespace std;
int precedence(char ch)
{
    if (ch == '^')
    {
        return 3;
    }
    else if (ch == '*' || ch == '/')
    {
        return 2;
    }
    else if (ch == '+' || ch == '-')
    {
        return 1;
    }
    else
    {
        return -1;
    }
}
string reverseString(string str)
{
    reverse(str.begin(), str.end());
    int l = str.size();
    for (int i = 0; i <= l; i++)
    {
        if (str[i] == ')')
        {
            str[i] = '(';
        }
        else if (str[i] == '(')
        {
            str[i] = ')';
        }
    }
    return str;
}

string infixToPostfix(string infixExpression)
{
    stack<char> st;
    int l = infixExpression.size();
    string postfixExpression = "";
    for (int i = 0; i <= l; i++)
    {
        int current = infixExpression[i];

        if (current >= 'a' && current <= 'z' || current >= 'A' && current <= 'Z')
        {
            postfixExpression += current;
        }
        else if (current == '(')
        {
            st.push(current);
        }
        else if (current == ')')
        {
            while (!st.empty() && st.top() != '(')
            {
                postfixExpression += st.top();
                st.pop();
            }
            st.pop();
        }
        else
        {
            while (!st.empty() && precedence(st.top()) >= precedence(current))
            {
                postfixExpression += st.top();
                st.pop();
            }
            st.push(current);
        }
    }
    while (!st.empty())
    {
        postfixExpression += st.top();
        st.pop();
    }
    return postfixExpression;
}
string infixToPrefix(string infixString)
{
    string infix = reverseString(infixString);
    string postfix = infixToPostfix(infix);
    string prefix = reverseString(postfix);
    return prefix;
}
int main()
{
    string infixExpression;
    char multipleTime = 'y';

    while (multipleTime == 'y')
    {
        cout << "Enter Infix Expression --> ";
        cin >> infixExpression;
        cout << "Prefix Expresion --> " << infixToPrefix(infixExpression) << endl;
        cout << "Want to decide (y/n) --> ";
        cin >> multipleTime;
    }
}

File: JAVA/Infix To Prefix/InfixToPrefix.java
import java.util.Scanner;
import java.util.Stack;


public class InfixToPrefix {
    public static String infixToPrefix(String infix)
    {
        String infixReverse = reverseString(infix);
        String postfix = infixToPostfix(infixReverse);
        String prefix = reverseString(postfix);
        return prefix;
    }

    public static String reverseString(String str)
    {
        String reverse = "";
        for (int i=str.length()-1; i>=0; i--)
        {
            char current = str.charAt(i);
            if (current == '(')
            {
                current = ')';
                i++;
            }
            else if (current == ')')
            {
                current = '(';
                i++; 
            }
            reverse += current;
        }
        return String.valueOf(reverse);
    }

    public static int precedence(char ch)
    {
        switch (ch)
        {
            case '+':
                return 1;
            case '-':
                return 1;
            case '*':
                return 2;
            case '/':
                return 2;
            case '^':
                return 3;
            default:
                return -1;
        }
    }
    
    public static String infixToPostfix(String infixExpression)
    {
        Stack<Character> st = new Stack<>();
        String postfixExpression = "";
        for (int i = 0; i < infixExpression.length(); i++) {
            char current = infixExpression.charAt(i);

            if ((current >= 'a' && current <= 'z') || (current >= 'A' && current <= 'Z')) {
                postfixExpression += current;
            }
            else if (current == '(')
            {
                st.push(current);
            }
            else if (current == ')')
            {
                while (!st.empty() && st.peek() != '(')
                {
                    postfixExpression += st.peek();
                    st.pop();
                }
                st.pop();
            }
            else
            {
                while (!st.empty() && precedence(st.peek()) >= precedence(current))
                {
                    postfixExpression += st.peek();
                    st.pop();
                }
                st.push(current);
            }
        }
        while (!st.empty())
        {
            postfixExpression += st.peek();
            st.pop();
        }
        return postfixExpression;
    }

    public static void main(String[] args) {
        String infixString;
        char multipleTime = 'y';

        while (multipleTime == 'y') {
            Scanner scanner = new Scanner(System.in);
            System.out.print("Enter Infix Expression --> ");
            infixString = scanner.nextLine();                
            System.out.println("Prefix Expression --> 2. " + infixToPrefix(infixString));
            System.out.print("Want to decide(y/n) --> ");
            multipleTime = scanner.next().charAt(0);
        }

    }
}
